hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/audit/AuditTool.java

```diff
@@ -0,0 +1,334 @@
+   **/
+  @VisibleForTesting
+  public static final String WRONG_FILESYSTEM = "Wrong filesystem for ";
+
+  private final String usage = entryPoint + "  s3a://BUCKET\n";
+
+  private final File s3aLogsDirectory = new File("S3AAuditLogsDirectory");
+
+  public AuditTool() {
+  }
+
+  /**
+   * tells us the usage of the AuditTool by commands.
+   *
+   * @return the string USAGE
+   */
+  public String getUsage() {
+    return usage;
+  }
+
+  /**
+   * this run method in AuditTool takes S3 bucket path.
+   * which contains audit log files from command line arguments
+   * and merge the audit log files present in that path into single file in local system
+   *
+   * @param args command specific arguments.
+   * @return SUCCESS i.e, '0', which is an exit code
+   * @throws Exception on any failure
+   */
+  @Override
+  public int run(String[] args) throws Exception {
+    List<String> argv = new ArrayList<>(Arrays.asList(args));
+    println("argv: %s", argv);
+    if (argv.isEmpty()) {
+      errorln(getUsage());
+      throw invalidArgs("No bucket specified");
+    }
+    //path of audit log files in s3 bucket
+    Path s3LogsPath = new Path(argv.get(0));
+
+    //setting the file system
+    URI fsURI = toUri(String.valueOf(s3LogsPath));
+    S3AFileSystem s3AFileSystem =
+        bindFilesystem(FileSystem.newInstance(fsURI, getConf()));
+    RemoteIterator<LocatedFileStatus> listOfS3LogFiles =
+        s3AFileSystem.listFiles(s3LogsPath, true);
+
+    //creating local audit log files directory and
+    //copying audit log files into local files from s3 bucket
+    //so that it will be easy for us to implement merging and parsing classes
+    if (!s3aLogsDirectory.exists()) {
+      boolean s3aLogsDirectoryCreation = s3aLogsDirectory.mkdir();
+    }
+    File s3aLogsSubDir = new File(s3aLogsDirectory, s3LogsPath.getName());
+    boolean s3aLogsSubDirCreation = false;
+    if (!s3aLogsSubDir.exists()) {
+      s3aLogsSubDirCreation = s3aLogsSubDir.mkdir();
+    }
+    if (s3aLogsSubDirCreation) {
+      while (listOfS3LogFiles.hasNext()) {
+        Path s3LogFilePath = listOfS3LogFiles.next().getPath();
+        File s3LogLocalFilePath =
+            new File(s3aLogsSubDir, s3LogFilePath.getName());
+        boolean localFileCreation = s3LogLocalFilePath.createNewFile();
+        if (localFileCreation) {
+          FileStatus fileStatus = s3AFileSystem.getFileStatus(s3LogFilePath);
+          long s3LogFileLength = fileStatus.getLen();
+          //reads s3 file data into byte buffer
+          byte[] s3LogDataBuffer =
+              readDataset(s3AFileSystem, s3LogFilePath, (int) s3LogFileLength);
+          //writes byte array into local file
+          FileUtils.writeByteArrayToFile(s3LogLocalFilePath, s3LogDataBuffer);
+        }
+      }
+    }
+
+    //calls S3AAuditLogMerger for implementing merging code
+    //by passing local audit log files directory which are copied from s3 bucket
```

Comment:
Again keep the comments simple, like merging local audit files into a single file.

Reply:
changed comment